contracts/sdk/indexer/db.ts

```diff
@@ -284,30 +336,139 @@ export class NFTDatabaseConnection {
     return leafIdxs;
   }
 
-  async getProof(hash: Buffer, treeId: string, check: boolean = true): Promise<Proof | null> {
+  async getMaxSeq(treeId: string) {
+    let res = await this.connection.get(
+      `
+        SELECT max(seq) as seq
+        FROM merkle 
+        WHERE tree_id = ?
+      `,
+      treeId
+    );
+    if (res) {
+      return res.seq;
+    } else {
+      return null;
+    }
+  }
+  async getInferredProof(
+    hash: Buffer,
+    treeId: string,
+    check: boolean = true
+  ): Promise<Proof | null> {
+    let latestSeq = await this.getMaxSeq(treeId);
+    if (!latestSeq) {
+      return null;
+    }
+    let gapIndex = await this.connection.get(
+      `
+        SELECT 
+          m0.seq as seq
+        FROM merkle m0
+        WHERE NOT EXISTS (
+          SELECT NULL
+          FROM merkle m1
+          WHERE m1.seq = m0.seq + 1 AND m1.tree_id = ?
+        ) AND tree_id = ?
+        ORDER BY m0.seq
+        LIMIT 1
+      `,
+      treeId,
+      treeId
+    );
+    if (gapIndex && gapIndex.seq < latestSeq) {
+      return await this.inferProof(hash, treeId, gapIndex.seq, check);
+    } else {
+      return await this.getProof(hash, treeId, check);
+    }
+  }
+
+  async inferProof(
+    hash: Buffer,
```

Comment:
i dislike this, would prefer to use leafIndex / nonce then map into a node_index